flutter_local_notifications/lib/src/platform_flutter_local_notifications.dart

```diff
@@ -202,6 +203,87 @@ class AndroidFlutterLocalNotificationsPlugin
+  ///  android:stopWithTask="true"
+  ///  android:foregroundServiceType="As you like" />
+  /// ```
+  /// While the `android:name` must exactly match this value, you can configure
+  /// the other parameters as you like, although it is recommended to copy the
+  /// values for `android:enabled`, `android:exported` and
+  /// `android:stopWithTask`. Suitable values for `foregroundServiceType` can be
+  /// found [here](https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification,%20int)).
+  ///
+  /// The notification of the foreground service can be updated by
+  /// simply calling this method multiple times.
+  ///
+  /// Information on selecting an appropriate `startType` for your app's usecase
+  /// should be taken from the official Android documentation, check [`Service.onStartCommand`](https://developer.android.com/reference/android/app/Service#onStartCommand(android.content.Intent,%20int,%20int)).
+  /// The there mentioned constants can be found in [AndroidServiceConstants].
+  ///
+  /// The notification for the foreground service will not be dismissable
+  /// and automatically removed when using [stopForegroundService].
+  ///
+  /// If `foregroundServiceType` is set, [`Service.startForeground(int id, Notification notification, int foregroundServiceType)`](https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification,%20int))
+  /// will be invoked , else  [`Service.startForeground(int id, Notification notification)`](https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification)) is used.
+  /// On devices older than [`Build.VERSION_CODES.Q`](https://developer.android.com/reference/android/os/Build.VERSION_CODES#Q), `foregroundServiceType` will be ignored.
+  /// Multiple type flags can be ORed together (using the `|` operator).
+  /// Note that `foregroundServiceType` (the parameter in this method)
+  /// must be a subset of the `android:foregroundServiceType`
+  /// defined in your `AndroidManifest.xml` (the one from the section above)!
+  Future<void> startForegroundService(int id, String? title, String? body,
+      {AndroidNotificationDetails? notificationDetails,
+      String? payload,
+      int startType = AndroidServiceConstants.startSticky,
+      int? foregroundServiceType}) {
+    validateId(id);
+    if (id == 0) {
+      throw ArgumentError.value(id, 'id',
+          'The id of a notification used for an Android foreground service must not be 0!'); // ignore: lines_longer_than_80_chars
+    }
+    return _channel.invokeMethod('startForegroundService', <String, Object>{
+      'notificationData': <String, Object?>{
+        'id': id,
+        'title': title,
+        'body': body,
+        'payload': payload ?? '',
+        'platformSpecifics': notificationDetails?.toMap(),
+      },
+      'startType': startType,
+      'hasForegroundServiceType': foregroundServiceType != null,
```

Comment:
This seems redundant as the checks could be done on the plugin side. This could also lead to confusion when debugging on the Java side as the `foregroundServiceType` would then be coalesced. This is highlighted by the fact that when a null `foregroundServiceType` is specified and then coalesced to 0, the value is never used. A value of 0 also corresponds to https://developer.android.com/reference/android/content/pm/ServiceInfo#FOREGROUND_SERVICE_TYPE_NONE. Not specifying any service types should mean to use those specified in the manifest file i.e. https://developer.android.com/reference/android/content/pm/ServiceInfo#FOREGROUND_SERVICE_TYPE_MANIFEST.

My suggestion here would to get remove `hasForegroundServiceType` and don't coalesce `foregroundServiceType`. This means passing in null would call https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification) whilst passing in a value would call https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification,%20int).

Note: bear in mind I have a comment further below in allowing the API to accept a list of foreground service types so it would mean a null list would call https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification)  whilst a non-empty list would call https://developer.android.com/reference/android/app/Service#startForeground(int,%20android.app.Notification,%20int). An empty list IMO is most likely a programmer error so an `assert` call could be made to verify that either a null or non-empty list is specified

Reply:
I agree on that.

About the last sentence of your comment: How would we proceed with double entries in that list? Ignore them (my preferred solution) or `assert` that every entry in the list is unique? If it wasn't for the syntactic sugar  of `[]` I would have suggested using a `Set` instead of a `List` in the method declaration...

Reply:
Ah yes `Set` is a better idea